doctr/models/recognition/sar/pytorch.py

```diff
@@ -93,65 +99,69 @@ def __init__(
-            (features.shape[0], self.max_length + 1, self.vocab_size + 1),
-            device=features.device,
-            dtype=features.dtype,
-        )
-
-        decoding_iter = self.max_length + 1 if gt is None else 1
-        for t in range(decoding_iter):
-            # (N, L + 1, rnn_units)
-            logits = self.rnn(symbol)[0]
-            # (N, L + 1, C)
-            glimpse = self.attention_module(features, logits)
-            # (N, L + 1, 2 * rnn_units)
-            logits = torch.cat((logits, glimpse), -1)
-            # (N, L + 1, vocab_size + 1)
-            decoded = self.output_dense(logits)
-            if gt is None:
-                char_logits[:, t] = decoded[:, t + 1]
-                if t < decoding_iter - 1:
-                    # update symbol with predicted logits for t + 1 step
-                    symbol[:, t + 2] = self.embed(decoded[:, t + 1].argmax(-1))
+        if gt is not None:
+            gt_embedding = self.embed_tgt(gt)
+
+        logits_list: List[torch.Tensor] = []
+
+        for t in range(self.max_length + 1):  # 32
+            if t == 0:
+                # step to init the first states of the LSTMCell
+                hidden_state_init = cell_state_init = \
+                    torch.zeros(features.size(0), features.size(1), device=features.device)
+                hidden_state, cell_state = hidden_state_init, cell_state_init
+                prev_symbol = holistic
+            elif t == 1:
+                # step to init a 'blank' sequence of length vocab_size + 1 filled with zeros
+                # (N, vocab_size + 1) --> (N, embedding_units)
+                prev_symbol = torch.zeros(features.size(0), self.vocab_size + 1, device=features.device)
+                prev_symbol = self.embed(prev_symbol)
             else:
-                char_logits = decoded[:, 1:]
-
-        return char_logits
+                if gt is not None:
+                    #(N, embedding_units) -2 because of <bos> and <eos> (same)
+                    prev_symbol = self.embed(gt_embedding[:, t - 2])
+                else:
+                    # -1 to start at timestep where prev_symbol was initialized
+                    index = logits_list[t - 1].argmax(-1)
+                    # (1, embedding_units)
+                    prev_symbol = prev_symbol.scatter_(1, index.unsqueeze(1), 1)
+
+            # (N, C), (N, C)  take the last hidden state and cell state from current timestep
+            hidden_state_init, cell_state_init = self.lstm_cell(prev_symbol, (hidden_state_init, cell_state_init))
+            hidden_state, cell_state = self.lstm_cell(hidden_state_init, (hidden_state, cell_state))
+            # (N, C, H, W), (N, C) --> (N, C)
+            glimpse = self.attention_module(features, hidden_state)
+            # (N, C), (N, C) --> (N, 2 * C)
+            logits = torch.cat([hidden_state, glimpse], dim=1)
+            if gt is not None:
+                logits = self.dropout(logits)
```

Comment:
Correct me if I'm wrong, but this is rather dangerous:
- here we apply dropout only when gt is not None 
- I would suggest to switch to a condition based on training mode `self.training` out of safety
- second, dropout only works if we use it both at train & test size : at test time, it rescales the distribution to match the situation observed during training


To make it short, I'd suggest:
```python
logits = self.dropout(logits)
```
without any condition (the layer inner forward is already making the proper checks)

Reply:
Oh good catch thanks :+1: